swupd/rename.go

```diff
@@ -0,0 +1,183 @@
+	// the manifest changes.
+	sort.Slice(added, func(i, j int) bool {
+		return added[i].Hash < added[j].Hash
+	})
+	sort.Slice(removed, func(i, j int) bool {
+		return removed[i].Hash < removed[j].Hash
+	})
+	for ax, rx := 0, 0; ax < len(added) && rx < len(removed); {
+		af := added[ax]
+		rf := removed[rx]
+		switch {
+		case af.Hash < rf.Hash:
+			ax++
+		case af.Hash > rf.Hash:
+			rx++
+		default: // Equal hash, so link
+			linkrenamepair(af, rf)
+			ax++
+			rx++
+		}
+	}
+	// Link things with the same names skippingdigits
+	// First remove small files (not worth sending diff) and files which have
+	// exact match
+	added = trimsmall(trimrenamed(added), BSDIFFSIZE)     // Make it explicit we are doing two steps
+	removed = trimsmall(trimrenamed(removed), BSDIFFSIZE) // Can trivially make it one pass.
+	//generate the pairs of *File and short name
+	pa := makepairs(added)
+	pr := makepairs(removed)
+	// Merge where short names match
+	for ax, rx := 0, 0; ax < len(pa) && rx < len(pr); {
+		af := pa[ax]
+		rf := pr[rx]
+		switch {
+		case af.n < rf.n:
+			ax++
+		case af.n > rf.n:
+			rx++
+		default: // Equal truncated name
+			linkrenamepair(af.f, rf.f)
+			ax++
+			rx++
+		}
+	}
+}
+
+// linkrenamepair links two files together
+func linkrenamepair(a, b *File) {
+	a.DeltaPeer = b
+	b.DeltaPeer = a
+	a.Rename = true
+	b.Rename = true
+}
+
+// trimrenamed returns an slice which has had files with DeltaPeers purged
+func trimrenamed(a []*File) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Status == StatusGhosted {
+			panic("Found ghosted file " + f.Name)
+		}
+		if f.DeltaPeer == nil {
+			r = append(r, f)
+		}
+	}
+	return r
+}
+
+// trimsmall returns an slice which has had files with small files removed
+func trimsmall(a []*File, minsize int64) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Info == nil {
+			panic("Missing f.Info for " + f.Name)
+		}
+		if f.Info.Size() > minsize {
+			r = append(r, f)
+		}
+	}
+	return r
+}
+
+// pairednames holds a *File and another "name", which is used for various
+// alterations of the filename for matching purposes. In particular removing
+// digits so a new release of a shared library is matched to previous ones
+type pairednames struct {
+	f *File
+	n string // Filename with digits removed
+}
+
+// stripdigits is used with strings.Map to remove digits from filename
+func stripdigits(r rune) rune {
+	if r >= '0' && r <= '9' {
+		return -1 // strings.Map removes negative values
+	}
+	return r
+}
+
+// makepairs returns a sorted by name, disregarding digits
+// secondary sort key is the original name. This is mainly inteneded for
+// cases like with python we have the same filename under both python3.6 and python2.7
+// it would look weird to rename the old python2.7/file1 to python3.6/file2 and at
+// the same time rename the old python3.6/file1 to python2.7/file2.
+func makepairs(a []*File) []pairednames {
```

Comment:
I'm not really understanding the `makepairs` name here since it doesn't pair any filenames here. Do you instead mean `makepares` and `parednames` or something since you are kind of paring the filenames?